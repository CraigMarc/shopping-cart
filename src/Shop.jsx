import { Header } from './Header'
import { useState, useEffect } from 'react'
import { Link } from "react-router-dom";

import image1 from './assets/images/image1.jpg';
import image2 from './assets/images/image2.jpg';
import image3 from './assets/images/image3.jpg';
import image4 from './assets/images/image4.jpg';
import image5 from './assets/images/image5.jpg';
import image6 from './assets/images/image6.jpg';
import image7 from './assets/images/image7.jpg';
import image8 from './assets/images/image8.jpg';
import image9 from './assets/images/image9.jpg';
import image10 from './assets/images/image10.jpg';
import image11 from './assets/images/image11.jpg';
import image12 from './assets/images/image12.jpg';


function Shop(props) {

  const {

    apiItems,
    setApiItems,
    cartItems

  } = props;

  // if api goes down use data in assets file uncomment and comment out fetch change setApiItems in app.jsx
  //<img className="img" src={imageArray[index.id - 1]}></img>

  //let imageArray = [image1, image2, image3, image4, image5, image6, image7, image8, image9, image10, image11, image12]



  if (apiItems == undefined) {
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(true);

    const fetchInfo = async (pics) => {
      //setLoading(true)

      try {
        //return fetch(picUrl)
        const res = await fetch("http://localhost:3000/users/all")
        // const res = await fetch("https://fakestoreapi.in/api/products")

        const productData = await res.json();
        console.log(productData)
        setApiItems(productData)

      }


      catch (error) {
        console.error("There has been a problem with your fetch operation:", error);
        //add error message to dom
        setError("true")

      }
      setLoading(false)

    }


    useEffect(() => {
      fetchInfo();
    }, [])

    //display error and loading for api call

    if (error) return (
      <div>
        <Header />
        <p>A network error was encountered</p>
      </div>
    )

    if (loading) return <p className="loading">Loading...</p>;

  }



  return (
    <div>
      <Header
        cartItems={cartItems}
      />
      <div className='shopContainer'>
        <h1>Our Products</h1>
        <div className="productCard">

          {apiItems.map((index, iter) => {

            let url = `http://localhost:3000/uploads/${index.image}`

            return (

              <div key={index.id} className="product">
                <Link to="/product" state={iter}>
                  <div id={index.id} className="card" >

                    <img className="img" src={url}></img>
                    <p>{index.title}</p>
                    <p className="price">${index.price}.00</p>

                  </div>
                </Link>
              </div>

            )
          })}

        </div>
      </div>
    </div>
  )


}

export default Shop