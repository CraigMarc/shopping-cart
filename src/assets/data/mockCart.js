let mockItems= [
    {
        "id": "638b5c8f-c2a6-4ba6-9984-75ea0bc26117",
        "title": "Mens Casual Premium Slim Fit T-Shirts ",
        "price": 22.3,
        "quantity": 1,
        "total": "22.30",
        "image": "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
    },
    {
        "id": "a4c6422a-0a8f-480d-a10c-6b8322b5d811",
        "title": "Mens Cotton Jacket",
        "price": 55.99,
        "quantity": "2",
        "total": "111.98",
        "image": "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"
    }
]

export default mockItems;