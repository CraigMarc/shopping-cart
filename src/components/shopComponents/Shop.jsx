import { Header } from '../headerComponents/Header'
import { Footer } from '../footerComponents/Footer'
import { HeaderMobile } from '../headerComponents/HeaderMobile'
import { useState, useEffect, useRef } from 'react'
import { Link, useLocation } from "react-router-dom";


function Shop(props) {

  const {

    apiItems,
    setApiItems,
    cartItems,
    category

  } = props;

  // scroll to top of page 

  useEffect(() => {
    window.scrollTo(0, 0)
  }, []);

  const location = useLocation();
  const pageData = location.state;
  const filteredProducts = useRef()

  // filter products

  const categoryData = apiItems.filter((product) => product.category._id == pageData.category._id)
  const brandData = apiItems.filter((product) => product.brand == pageData.category)


  if (categoryData.length == 0 && brandData.length == 0) {
    filteredProducts.current = null
  }

  if (categoryData.length > 0) {
    filteredProducts.current = categoryData
  }

  if (pageData.subCategory) {
    filteredProducts.current = categoryData.filter((product) => product.subCategory == pageData.subCategory)
    
  }

  if (brandData.length > 0) {
    filteredProducts.current = brandData
   
  }


  if (pageData.category._id == "all") {
    let newArray = structuredClone(apiItems)
    filteredProducts.current = newArray
   
  }

  if (pageData.category._id == "sale") {
    const saleData = apiItems.filter((product) => product.sale_percent > 0)
    filteredProducts.current = saleData
   
  }

  if (pageData.category._id == "search") {
    let lowerCase = pageData.category.name.toLowerCase()
    

    if (lowerCase.charAt(lowerCase.length - 1) == 's') {
     let nonPlural = lowerCase.substring(0, lowerCase.length - 1)
      let searchData = apiItems.filter((product) => product.title.toLowerCase().includes(nonPlural) || product.description.toLowerCase().includes(nonPlural) || product.category.name.toLowerCase().includes(nonPlural) || product.title.toLowerCase().includes(lowerCase) || product.description.toLowerCase().includes(lowerCase) || product.category.name.toLowerCase().includes(lowerCase))
      filteredProducts.current = searchData
    }

    if (lowerCase.charAt(lowerCase.length - 1) != 's') {
      let plural = lowerCase + 's'
      let searchData = apiItems.filter((product) => product.title.toLowerCase().includes(plural) || product.description.toLowerCase().includes(plural) || product.category.name.toLowerCase().includes(plural) || product.title.toLowerCase().includes(lowerCase) || product.description.toLowerCase().includes(lowerCase) || product.category.name.toLowerCase().includes(lowerCase))
      filteredProducts.current = searchData
    }
   
  }

  if (!filteredProducts.current) {
    filteredProducts.current = []
  }

  // display sale price or reg price

  function RenderSale(props) {

    const {

      index

    } = props

    let priceDiv = (index.colorArray[0].sizeArray[0].price / 100).toFixed(2)
    let salePrice = (priceDiv - (priceDiv * (index.sale_percent / 100))).toFixed(2)

    if (index.sale_percent == null) {
      return (
        <p className="price">${priceDiv}</p>
      )
    }

    else {
      return (
        <div>
          <p className='salePercentShop'>you save {index.sale_percent}%</p>
          <p className='regPrice'>${priceDiv}</p>
          <p className='price'>${salePrice}</p>
        </div>
      )
    }

  }

  function RenderTitle() {
    if (pageData.subCategory) {
      return (
        <h1>{pageData.subCategory}</h1>
      )
    }
    if (!pageData.category.name) {
      return (
        <h1>{pageData.category}</h1>
      )
    }
    else {
      return (
        <h1>{pageData.category.name}</h1>
      )
    }
  }

   

  // display products if exist

  function RenderProducts() {

    if (filteredProducts.current.length > 0) {
      return (
        <div className="productCard">
          {filteredProducts.current.map((index, iter) => {

            let url = ""
            if (index.colorArray[0].images) {
              let image = index.colorArray[0].images[0]
              url = `https://shoppingapi.fly.dev/${image}`
            }



            return (

              <Link key={iter} to="/product" state={index._id}>
                <div className="product">

                  <div id={index.id} className="card" >

                    <img className="img" alt="no image available" src={url}></img>
                    <p>{index.title}</p>
                    <RenderSale
                      index={index}
                    />

                  </div>

                </div>
              </Link>
            )
          })}

        </div>
      )
    }
    else {
      return (
        <div>
          <h3>There are no products currently available.</h3>
        </div>
      )
    }

  }

  return (
    <div>
      {window.innerWidth > 630 ? <Header cartItems={cartItems}
        category={category} /> : <HeaderMobile cartItems={cartItems}
          category={category} />}
      <div className='shopContainer'>
        <RenderTitle />

        <RenderProducts />

      </div>
      <Footer />
    </div>
  )


}

export default Shop