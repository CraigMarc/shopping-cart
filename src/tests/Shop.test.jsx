import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
//import TestApp from "./src/TestApp";
import { describe, it, expect } from 'vitest';
import Shop from "../Shop"
import data from "../assets/data/data.json"

import { BrowserRouter, Routes, Route, useParams } from 'react-router-dom';

//test1
it('links to a product page', async () => {
   
    render(
        <BrowserRouter >
            <Shop
            apiItems={data} />
             <Routes>
          
          <Route path="/product" element={<h1>Product</h1>} />
        </Routes>
          
        </BrowserRouter>
    );

    const user = userEvent.setup();

    const links = screen.getByText('Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops');
   
    await user.click(links)
    
    
    
   expect(
        screen.getByRole('heading', { level: 1, name: 'Product' }),
    ).toBeInTheDocument();
});

// test 2

it("renders home page", () => {
    // since screen does not have the container property, we'll destructure render to obtain a container for this test
    const { container } = render(
        <BrowserRouter >
        <Shop
        apiItems={data}
        
        />
       
      
    </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  


