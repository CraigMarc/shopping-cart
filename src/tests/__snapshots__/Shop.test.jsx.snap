// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders home page 1`] = `
<div>
  <div>
    <div
      class="headerDiv"
    >
      <header>
        <h1>
          My Store
        </h1>
        <a
          class="heading"
          href="/"
        >
          <div>
            Home
          </div>
        </a>
        <a
          class="heading"
          href="/cart"
        >
          <div
            class="cartContainer"
          >
            <span
              class="itemNumber"
            >
              0
            </span>
            <img
              class="shoppingCart"
              src="/src/assets/shoppingCart.png"
            />
          </div>
        </a>
      </header>
    </div>
    <div
      class="shopContainer"
    >
      <h1>
        Our Products
      </h1>
      <div
        class="productCard"
      >
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="1"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
              />
              <p>
                Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
              </p>
              <p>
                $
                109.95
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="2"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
              />
              <p>
                Mens Casual Premium Slim Fit T-Shirts 
              </p>
              <p>
                $
                22.3
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="3"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"
              />
              <p>
                Mens Cotton Jacket
              </p>
              <p>
                $
                55.99
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="4"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
              />
              <p>
                Mens Casual Slim Fit
              </p>
              <p>
                $
                15.99
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="5"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg"
              />
              <p>
                John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet
              </p>
              <p>
                $
                695
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="6"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/61sbMiUnoGL._AC_UL640_QL65_ML3_.jpg"
              />
              <p>
                Solid Gold Petite Micropave 
              </p>
              <p>
                $
                168
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="7"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71YAIFU48IL._AC_UL640_QL65_ML3_.jpg"
              />
              <p>
                White Gold Plated Princess
              </p>
              <p>
                $
                9.99
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="8"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/51UDEzMJVpL._AC_UL640_QL65_ML3_.jpg"
              />
              <p>
                Pierced Owl Rose Gold Plated Stainless Steel Double
              </p>
              <p>
                $
                10.99
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="9"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
              />
              <p>
                WD 2TB Elements Portable External Hard Drive - USB 3.0 
              </p>
              <p>
                $
                64
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="10"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg"
              />
              <p>
                SanDisk SSD PLUS 1TB Internal SSD - SATA III 6 Gb/s
              </p>
              <p>
                $
                109
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="11"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg"
              />
              <p>
                Silicon Power 256GB SSD 3D NAND A55 SLC Cache Performance Boost SATA III 2.5
              </p>
              <p>
                $
                109
              </p>
            </div>
          </a>
        </div>
        <div
          class="product"
        >
          <a
            href="/product"
          >
            <div
              class="card"
              id="12"
            >
              <img
                class="img"
                src="https://fakestoreapi.com/img/61mtL65D4cL._AC_SX679_.jpg"
              />
              <p>
                WD 4TB Gaming Drive Works with Playstation 4 Portable External Hard Drive
              </p>
              <p>
                $
                114
              </p>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
